cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java

```diff
@@ -312,11 +311,24 @@ private HttpResponse doRequest(HttpMethod method, String resource,
     }
   }
 
-  private HttpRequest.Builder processBuilder(HttpRequest.Builder builder) {
-    if (SecurityRequestContext.getUserId() != null) {
-      builder.addHeader(Constants.Security.Headers.USER_ID, SecurityRequestContext.getUserId());
+  private HttpRequest.Builder addUserIdHeader(HttpRequest.Builder builder) throws IOException {
+    if (!securityEnabled) {
+      return builder;
+    }
+    // If the request originated from the router and was forwarded to any service other than dataset service, before
+    // going to dataset service via dataset service client, the userId could be set in the SecurityRequestContext.
+    // e.g. deploying an app that contains a dataset
+    String userId = SecurityRequestContext.getUserId();
+    if (NamespaceId.SYSTEM.equals(namespaceId.toEntityId())) {
+      // For getting a system dataset like MDS, use the system principal. It is ok to do so, since DatasetServiceClient
+      // is an internal client that is not exposed to users.
```

Comment:
I've reported https://issues.cask.co/browse/CDAP-6583. Will come up with a design and fix this separately.